src/modm/platform/spi/rp/spi_master.hpp.in

```diff
@@ -0,0 +1,201 @@
+	static inline void *context{nullptr};
+	static inline ConfigurationHandler configuration{nullptr};
+protected:
+	static inline bool isBusy() {
+		return (spi{{ id }}_hw->sr & SPI_SSPSR_BSY_BITS);
+	}
+
+	static inline bool txFifoFull() {
+		return (spi{{ id }}_hw->sr & SPI_SSPSR_TNF_BITS) == 0;
+	}
+	static inline bool txFifoEmpty() {
+		return !!(spi{{ id }}_hw->sr & SPI_SSPSR_TFE_BITS);
+	}
+
+	static inline bool rxFifoEmpty() {
+		return (spi{{ id }}_hw->sr & SPI_SSPSR_RNE_BITS) == 0;
+	}
+
+	static inline void drainRx() {
+		// Drain RX FIFO, then wait for shifting to finish (which may be *after*
+	    // TX FIFO drains), then drain RX FIFO again
+	    while (!rxFifoEmpty())
+	        (void)read();
+	    while (isBusy())
+	        __NOP();
+	    while (!rxFifoEmpty())
+	        (void)read();;
+
+	    // Don't leave overrun flag set
+	    spi{{ id }}_hw->icr = SPI_SSPICR_RORIC_BITS;
+	}
+public:
+	template< class... Signals >
+	static void
+	connect()
+	{
+		using Connector = GpioConnector<Peripheral::Spi{{ id }}, Signals...>;
+		using Sclk = typename Connector::template GetSignal< PeripheralPin::Sclk >;
+		using Mosi = typename Connector::template GetSignal< PeripheralPin::Tx >;
+		/*using Miso = typename Connector::template GetSignal< PeripheralPin::Rx >;*/
+		static_assert(Connector::template IsValid<Sclk>);
+		static_assert(Connector::template IsValid<Mosi>);
+
+		Connector::connect();
+	}
+
+	static void reset();
+	static void unreset();
+
+	static void setFormat(uint8_t data_bits,DataMode mode,DataOrder order) {
+		modm_assert(data_bits >= 4 && data_bits <= 16, "SPI data_bits","SPI data bits out of range");
+		// LSB-first not supported on PL022:
+    	modm_assert(order == DataOrder::MsbFirst, "SPI DataOrder", "SPI hardware does not support alternate transmit order");
+		hw_write_masked(&spi{{ id }}_hw->cr0,
+                    ((uint32_t(data_bits - 1)) << SPI_SSPCR0_DSS_LSB) |
+                    (((uint32_t(mode)&2)>>1) << SPI_SSPCR0_SPO_LSB) |
+                    ((uint32_t(mode)&1) << SPI_SSPCR0_SPH_LSB),
+			        SPI_SSPCR0_DSS_BITS |
+			        SPI_SSPCR0_SPO_BITS |
+			        SPI_SSPCR0_SPH_BITS);
+	}
+
+	template< class SystemClock, baudrate_t baudrate >
+	static uint32_t setBaudrate() {
+		constexpr uint32_t freq_in = SystemClock::PeriFrequency;
+		uint32_t prescale,postdiv;
+		// Find smallest prescale value which puts output frequency in range of
+	    // post-divide. Prescale is an even number from 2 to 254 inclusive.
+	    for (prescale = 2; prescale <= 254; prescale += 2) {
+	        if (freq_in < (prescale + 2) * 256 * (uint64_t) baudrate)
+	            break;
+	    }
+	    //atatic_assert(prescale <= 254); // Frequency too low
+
+	    // Find largest post-divide which makes output <= baudrate. Post-divide is
+	    // an integer in the range 1 to 256 inclusive.
+	    for (postdiv = 256; postdiv > 1; --postdiv) {
+	        if (freq_in / (prescale * (postdiv - 1)) > baudrate)
+	            break;
+	    }
```

Comment:
Are any of the constexpr prescaler algorithms useful for this?
https://github.com/modm-io/modm/blob/develop/src/modm/math/algorithm/prescaler.hpp